### PR TITLE
Docs: Initial documentation for current JS architecture

### PR DESCRIPTION
**Pull request by gtrufitt**
## What does this change?

Adds a document describing the current architecture of our Javascript.

@guardian/dotcom-platform

### CODE REVIEW

File: docs/02-architecture/06-client-side-architecture.md [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+				- app.js
+				- commercial.js
+				- enhanced/main.js
+			- Cloudwatch beacon
+		- [inlineJSNonBlocking.scala.html](https://github.com/guardian/frontend/blob/master/common/app/views/fragments/inlineJSNonBlocking.scala.html)
+			- getUserData.js
+			- detectAdblock 
+			- showUserName
+			- ophanConfig
+
```
Comment (dominickendrick) [inline docs/02-architecture/06-client-side-architecture.md]: Should we also mention [analytics loading](https://github.com/guardian/frontend/blob/master/common/app/views/main.scala.html#L112) here?

File: docs/02-architecture/06-client-side-architecture.md [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+		
+#### Curl config
+
+Contains the [config](https://github.com/guardian/frontend/blob/master/common/app/templates/inlineJS/blocking/curlConfig.scala.js) for the [curl AMD module loader](https://github.com/cujojs/curl). You can find the aliases for certain JS modules in here such as `fastdom` which helps you avoid having to add the full path for certain modules.
+
+#### shouldEnhance
+
+The [shouldEnhance JS](https://github.com/guardian/frontend/blob/master/common/app/templates/inlineJS/blocking/shouldEnhance.scala.js) defines whether we should run the enhanced Javascript. 
+
+There are ways to force enhancement, so take a look at the JS, but something of note is that we don't enhance devices running iOS < 8 on all pages or on all iPads on Fronts.
```
Comment (dominickendrick) [inline docs/02-architecture/06-client-side-architecture.md]: Could this
> we don't enhance devices running iOS < 8 on all pages or on all iPads on Fronts. 

change to

we don't enhance devices running iOS < 8 on any pages and don't enhance iPads on Fronts.

File: docs/02-architecture/06-client-side-architecture.md [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+
+[Puts the user's username](https://github.com/guardian/frontend/blob/master/common/app/templates/inlineJS/nonBlocking/showUserName.scala.js) into the header (increases perceived rendering speed as their username is there before the app.js has to download and parse).
+
+#### Ophan config
+
+Gets the [Ophan browserId](https://github.com/guardian/frontend/blob/master/common/app/templates/inlineJS/nonBlocking/ophanConfig.scala.js) which is used across analytics to tie data together.
+
+### Bootstraps
+
+In javascripts/bootstraps we define all the entry points for each bundle described in [requirejs.js]().
```
Comment (dominickendrick) [inline docs/02-architecture/06-client-side-architecture.md]: can we code format `javascripts/bootstraps` and link to the file ?

File: docs/02-architecture/06-client-side-architecture.md [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+And then specific Javascript bootstraps used on different pages and required into the app when needed - Article, Article Minute, Crosswords, Liveblog, Gallery, Trail, Profile, Sudoku, Image content, Facia, Football, Preferences, Membership, Ophan, Admin, Main Media, Video Embed, Accessibility.
+
+### General structure
+
+See below for quick descriptions.
+
+- [- main.scala.html](https://github.com/guardian/frontend/blob/master/common/app/views/main.scala.html)
+	- [head.scala.html](https://github.com/guardian/frontend/blob/master/common/app/views/fragments/head.scala.html)
+		- [inlineJsBlocking.scala.html](https://github.com/guardian/frontend/blob/master/common/app/views/fragments/inlineJSBlocking.scala.html) - Scripts required to render the page correctly on first paint.
+			- Polyfills for json2, es5-html5, RAF, classlist
```
Comment (SiAdcock) [inline docs/02-architecture/06-client-side-architecture.md]: Also polyfill for `<details>` 😉

File: docs/02-architecture/06-client-side-architecture.md [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+The [shouldEnhance JS](https://github.com/guardian/frontend/blob/master/common/app/templates/inlineJS/blocking/shouldEnhance.scala.js) defines whether we should run the enhanced Javascript. 
+
+There are ways to force enhancement, so take a look at the JS, but something of note is that we don't enhance devices running iOS < 8 on all pages or on all iPads on Fronts.
+
+#### pageConfig
+
+If you put `guardian.config` in your console, you will see the JS config containing information about analytics, modules, switches, tests and the page. This config object is initially populated serverside and topped-up client side (i.e.: information about the user is only available client-side)
+
+The initial config structure is defined in [config.scala.js](https://github.com/guardian/frontend/blob/master/common/app/templates/inlineJS/blocking/config.scala.js) and [javaScriptConfig.scala.js](https://github.com/guardian/frontend/blob/master/common/app/templates/javaScriptConfig.scala.js). 
+
```
Comment (SiAdcock) [inline docs/02-architecture/06-client-side-architecture.md]: Is it worth mentioning here that `config.scala.js` also sets the `isModernBrowser` flag, which determines whether a user's browser cuts the mustard? Assuming this is actually used anywhere 😄

File: docs/02-architecture/06-client-side-architecture.md [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+You'll most likely be using the page config the most often which is defined in [JavascriptPage.scala](https://github.com/guardian/frontend/blob/master/common/app/views/support/JavaScriptPage.scala). Be aware that the metadata for a particular page may be [overridden with MetaData.make](https://github.com/guardian/frontend/blob/master/common/app/common/commercial/hosted/HostedGalleryPage.scala#L35).
+
+*Note that if you want to use config in a JS module, you shouldn't use the window object, but include it directly in your module via the config.js utility.*
```
Comment (SiAdcock) [inline docs/02-architecture/06-client-side-architecture.md]: Link to the `config.js` utility?

File: docs/02-architecture/06-client-side-architecture.md [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+app: {
+	src: [
+		options.staticSrcDir + 'javascripts/components/curl/curl-domReady.js',
+		options.staticTargetDir + 'javascripts/boot.js'
+	],
+	dest: options.staticTargetDir + 'javascripts/app.js'
+}
+```
+
+
```
Comment (SiAdcock) [inline docs/02-architecture/06-client-side-architecture.md]: Need (placeholder for?) `editionaliseMenu` function (part of the new header navigation test). Shows and hides menu items in the navigation depending on which edition of the site is currently selected

File: docs/02-architecture/06-client-side-architecture.md (hunk 3 above)
Comment (gtrufitt) [inline docs/02-architecture/06-client-side-architecture.md]: Ah yeah, whoops, did the last bit of this on the train and forgot I needed to link a few things!

File: docs/02-architecture/06-client-side-architecture.md [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+> // This file is intended to be downloaded and run ASAP on all pages by all readers.
+
+> // While it's ok to run code from here that requires specific host capabilities, it should manage failing gracefully by itself.
+
+>// Assume *nothing* about the host...
+
+> // This also means you should think *very hard* before adding modules to it, in particular 3rd party modules.
+
+> // For this file, performance and breadth of support should take priority over *anything*…
```
Comment (SiAdcock) [inline docs/02-architecture/06-client-side-architecture.md]: Style thing: I don't think the above needs to be presented as comments, it seems unnecessary

File: docs/02-architecture/06-client-side-architecture.md [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+> Bundles we need to run: commercial + enhanced
+
+> Only if we detect we should run enhance.
+
+It uses promises to require and init, in blocking order, standard JS, commercial JS and enhanced JS. As mentioned above it is bundled into app.js and the `insertRequire` option is used to insert a require call for it.
+
+
+### Commercial JS
+
+The commercial JS is it's own bundle and is executed immediately after the standard JS.
```
Comment (SiAdcock) [inline docs/02-architecture/06-client-side-architecture.md]: it's -> its

**Comment by SiAdcock** [comment]
Awesome docs, dude! 👍 x1000

File: docs/02-architecture/06-client-side-architecture.md [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+
+### Components
+
+[Here be 3rd-party JS](https://github.com/guardian/frontend/tree/master/static/src/javascripts/components) such as bean, bonzo and fastdom.
+
+### Projects
+
+There are five projects in the Javascript architecture:
+
+- Admin - This is the Javascript for the admin tools at frontend.gutools (not guardian.com related)
```
Comment (dominickendrick) [inline docs/02-architecture/06-client-side-architecture.md]: can `frontend.gutools` be a link ?

**Comment by desbo** [comment]
One thought I have on this is that it seems incredibly tightly coupled to the code, so it creates a maintenance cost of keeping this document up to date with any code changes. Maybe the content here would make more sense as comments in the appropriate places, so it's more obvious that it needs to be kept in line with the code. I don't know. Is this fixing the symptoms of a confusing architecture rather than addressing the root cause?

**Comment by gtrufitt** [comment]
@desbo Yeah, totally agree on the tight coupling and this being a symptom.

This was supposed to be more of  a getting started guide but it's difficult to write at that level with the complexity of our app - I think it's still high-level enough as a document for understanding the architecture as it doesn't go too deeply into the code *but* I will add links to this document from the code so it becomes the canonical reference to the high-level architecture.

If we're modifying the architecture of our app, we should really be updating the documentation to help others (that's gonna be the smallest part of that update task though right?).

One of my main personal reasons for writing the documentation is to enable me to finally get a full understanding of how and where all our JS is loaded, why and when and it's helped me look at it with a more analytical mind with a view to the future architecture. In order to fight the causes of complexity, we need to understand them and the symptoms are the entry point for that understanding.

I think we should get a WG together to fully understand *why* we have this complexity and techniques for removing it (because I think right now talks of webpack, rollup etc are going to introduce more), but for a while now we're going to have this architecture and I think seasoned and new developers alike coming to it need some kind of docs in one place.

File: docs/02-architecture/06-client-side-architecture.md (hunk 5 above)
Comment (gtrufitt) [inline docs/02-architecture/06-client-side-architecture.md]: Commercial JS doesn't run if `isModernBrowser` is false 😱

**Comment by desbo** [comment]
>  I will add links to this document from the code so it becomes the canonical reference to the high-level architecture

That's not what I meant – that'll only make it more tightly coupled.

> If we're modifying the architecture of our app, we should really be updating the documentation to help others (that's gonna be the smallest part of that update task though right?).

External documentation isn't a necessity. With the right architecture, the code *is* the documentation. This kind of document is basically a paraphrasing of the code in English that's expensive to keep up to date. If the changes we make are such that we have to update some external document to explain them, they're probably not the right changes to make.

**Comment by gtrufitt** [comment]
I'm not entirely sure whether you're saying this PR shouldn't be merged or whether we agree that right now we don't have a perfect world and this is sadly a needed document that we will hopefully not need in the future?

**Comment by desbo** [comment]
Sorry, I should have said I'm not saying this to stop it being merged, just to raise some questions about this kind of documentation. I'm not convinced it's needed, but if it turns out to be useful (do we have a way of measuring that?) and not too bad to maintain, I'd be happy to be proven wrong!

**Comment by gtrufitt** [comment]
Yeah, sure, we can always review it at a later point. Personally I think it's needed and a lot quicker to get a high-level understanding than clicking through files but I definitely agree that's a symptom of a bigger problem...

**Pull request merged**